staging/src/k8s.io/client-go/tools/cache/delta_fifo.go

```diff
@@ -526,6 +530,7 @@ func (f *DeltaFIFO) Pop(process PopProcessFunc) (interface{}, error) {
 
 			f.cond.Wait()
 		}
+		isInInitialList := !f.hasSynced_locked()
```

Comment:
My understanding is `isInInitialList` means "this is from the initial list".

The issue I have as a consumer is that I still can't tell when the initial list is _done_. I can't wait for an Add with `isInInitialList=false`, because there may never be one.

I would have expected this to be `isLastItemInList := f.hasSynced_locked()` (at L539, not here). Then a consumer can handle in their add controller `if isLastItemInList { do what I want on sync }`.

This may also lean towards having an `OnInitiallListCompleteFunc()` callback rather than a bool in the `AddFunc`, but that is stylistic - I don't have a strong preference

Reply:
What you describe won't work because eventually people start doing work async, knowing the last one isn't helpful, you need to know every one that's still in flight.

I will show how to use this in at least one controller.

The final check is going to look like `informer.HasSynced() && !initialListItems.InFlight()`, the latter part isn't in this PR yet.

Reply:
My thinking was that each layer needs to propagate some notion of "done". Anytime you add async, you are breaking that, and need to add it back at that layer. The goal of this is to propogate it at the informer handler layer. If you have _another_ layer on top of that (multiple worker queues, in k8s typically), then you will need some higher level solution on top of that which can be done by just knowing once all the event handlers have been called. _how_ exactly to do that depends on the code on top of the informer.

Its a bit hard to discuss the final check since I may be misunderstanding the intended flow, but I am guessing it is keeping track of all the items that come in with isInInitialList=true and then removing them as the queue completes?
 
If so, IMO it would still be preferable to flip the logic. 

If users are using multi worker queues, they _can_ still do this by just keeping track of items that come in before OnInitiallListCompleteFunc triggers. But for users who don't have this setup, they shouldn't have to pay the complexity of tracking items when a simple callback will do IMO.

Reply:
Users that aren't doing anything async already don't have any problems, so I don't understand why I shouldn't optimize it for people that are doing async things?

Reply:
(BTW I agree having a separate function to call on completion of the initial list would be better if we were starting from scratch, but I think what I've done here is going to have more minimal changes for consumers)

Reply:
This might just be a terminology thing - but I am not sure my understanding aligns with yours on non-async being ok as is.

When I am referring to non-async handlers, I mean something like this: https://github.com/kubernetes/kubernetes/blob/57eb5d631ccd615cd161b6da36afc759af004b93/staging/src/k8s.io/apiserver/pkg/admission/configuration/mutating_webhook_manager.go#L100, while an async handler is typically `queue.Add`.

With that definition, non-sync users are still broken - they still cannot tell when they are fully synced because the informer internals are also async and don't expose a way to know when the initial list is complete (leading to https://github.com/kubernetes/kubernetes/issues/113761).

> more minimal changes for consumers

My initial thought was just adding a field to `ResourceEventHandlerFuncs`; adding a field seems like its backwards compatible in almost all cases except the cases of keyless struct initialization, which seems rare. But I didn't dig into it much, so maybe there is more issues?

---

To be clear - I think this works and I am far happier to have this than nothing, consider these nitpicks :slightly_smiling_face:

Reply:
.... that code does N*log(N) work on start up for N webhooks, that's really bad, wow.

It appears to be wrong because HasSynced is called on the shared part of the informer, but not all notifications have made it through to the non-shared stub? That seems like a separate problem that I can also fix.

Reply:
> this controller doesn't care about the underlying store since it maintains its own index on top of things

Specifically the problem is the lag -- shared informer pipes everything through a channel, HasSynced is correct except for the stuff sitting in that channel. If this were using a non-shared informer it'd actually be correct; in that case HasSynced won't return true until after your Add handler for the last initial item has returned.

Reply:
That doesn't align with my experience. Both shared and unshared informers suffer the same issue. The issue comes down to `processorListener` which is used by all informers; `processorListener.add()` is async. If it wasn't, then everything would work fine actually (that is, `HasSynced` would also mean "event handlers have seen initial state", not just the store). This is why "Option 1: Make processorListener synchronous" in https://github.com/kubernetes/kubernetes/issues/113763 resolves the issue (although its probably a bad idea)

Here is a test of both shared and unshared that can reproduce it. Note its flaky - fails about 1% of the timer on my machine, so you need to run it with `stress` or similar.

<details>

```go
package main

import (
	"fmt"
	"sync/atomic"
	"testing"
	"time"

	v1 "k8s.io/api/core/v1"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/runtime"
	"k8s.io/apimachinery/pkg/util/wait"
	"k8s.io/client-go/informers"
	informersv1 "k8s.io/client-go/informers/core/v1"
	"k8s.io/client-go/kubernetes/fake"
	"k8s.io/client-go/tools/cache"
)

func TestInformerUnshared(t *testing.T) {
	kc := fake.NewSimpleClientset(pods...)
	handled := &atomic.Int32{}
	podInformer := informersv1.NewFilteredPodInformer(kc, "", 0, cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc}, nil)
	podInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
		AddFunc: func(obj any) {
			handled.Add(1)
		},
	})

	stop := make(chan struct{})
	go podInformer.Run(stop)
	WaitForCacheSync(stop, podInformer.HasSynced)

	if h := handled.Load(); h != int32(len(pods)) {
		t.Fatalf("got %+v events, expected %v", h, len(pods))
	}
}

func TestInformerShared(t *testing.T) {
	kc := fake.NewSimpleClientset(pods...)
	informers := informers.NewSharedInformerFactory(kc, 0)
	handled := &atomic.Int32{}
	podInformer := informers.Core().V1().Pods().Informer()
	podInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
		AddFunc: func(obj any) {
			handled.Add(1)
		},
	})

	stop := make(chan struct{})
	informers.Start(stop)
	WaitForInformerFactorySync(stop, informers)
	WaitForCacheSync(stop, podInformer.HasSynced)

	if h := handled.Load(); h != int32(len(pods)) {
		t.Fatalf("got %+v events, expected %v", h, len(pods))
	}
}

func WaitForInformerFactorySync(stop <-chan struct{}, informerFactory informers.SharedInformerFactory) {
	returnImmediately := make(chan struct{})
	close(returnImmediately)
	_ = wait.PollImmediate(time.Microsecond*10, wait.ForeverTestTimeout, func() (bool, error) {
		select {
		case <-stop:
			return false, fmt.Errorf("channel closed")
		default:
		}
		for _, synced := range informerFactory.WaitForCacheSync(returnImmediately) {
			if !synced {
				return false, nil
			}
		}
		return true, nil
	})
}

// Like standard WaitForCacheSync but more aggressive, to more easily detect race conditions.
func WaitForCacheSync(stop <-chan struct{}, cacheSyncs ...cache.InformerSynced) bool {
	max := time.Millisecond * 100
	delay := time.Millisecond
	f := func() bool {
		for _, syncFunc := range cacheSyncs {
			if !syncFunc() {
				return false
			}
		}
		return true
	}
	for {
		select {
		case <-stop:
			return false
		default:
		}
		res := f()
		if res {
			return true
		}
		delay *= 2
		if delay > max {
			delay = max
		}
		time.Sleep(delay)
	}
}

var pods = []runtime.Object{
	&v1.Pod{
		ObjectMeta: metav1.ObjectMeta{
			Name:      "a",
			Namespace: "n",
		},
		Spec: v1.PodSpec{},
	},
	&v1.Pod{
		ObjectMeta: metav1.ObjectMeta{
			Name:      "b",
			Namespace: "n",
		},
		Spec: v1.PodSpec{},
	},
	&v1.Pod{
		ObjectMeta: metav1.ObjectMeta{
			Name:      "c",
			Namespace: "n",
		},
		Spec: v1.PodSpec{},
	},
	&v1.Pod{
		ObjectMeta: metav1.ObjectMeta{
			Name:      "d",
			Namespace: "n",
		},
		Spec: v1.PodSpec{},
	},
}
```

</details>

Reply:
NewFilteredPodInformer still makes a shared informer: https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/staging/src/k8s.io/client-go/informers/core/v1/pod.go#L58-L59

Reply:
Ahhh... thank you. That explains a lot of confusion I had even beyond this issue. Thanks!

So the "Option 1: Make processorListener synchronous" is actually not _that_ crazy - that is exactly how the non-shared variant works anyways. It still may not be ideal since now unrelated handlers can block each other though (not a concern for my personal use cases, but I can see how that would be for some users).

Reply:
Well, we can't let one bad handler block others, so that's not feasible. I'll figure out a fix here in a bit.

Reply:
OK, check out what that controller looks like now.

🙏 that tests pass still ;)

(I need to add some tests)